apps/arweave/src/ar_merkle.erl

```diff
@@ -208,81 +304,96 @@ extract_root(_) ->
 %%% Private functions.
 %%%===================================================================
 
-generate_leaves(Elements) ->
-	lists:foldr(
-		fun({Data, Note}, Nodes) ->
-			Hash = hash([hash(Data), hash(note_to_binary(Note))]),
-			insert(
-				#node {
-					id = Hash,
-					type = leaf,
-					data = Data,
-					note = Note,
-					max = Note
-				},
-				Nodes
-			)
-		end,
-		new(),
-		Elements
-	).
+generate_tree([Element | Elements], Stack, Tree) when is_list(Element) ->
+	{SubRoot, SubTree} = generate_tree(Element),
+	SubTree2 = [mark_rebased(Node, SubRoot) || Node <- SubTree],
+	SubRootN = get(SubRoot, SubTree2),
+	generate_tree(Elements, queue:in(SubRootN, Stack), Tree ++ SubTree2);
+generate_tree([Element | Elements], Stack, Tree) ->
+	Leaf = generate_leaf(Element),
+	generate_tree(Elements, queue:in(Leaf, Stack), [Leaf | Tree]);
+generate_tree([], Stack, Tree) ->
+	case queue:to_list(Stack) of
+		[] ->
+			{<<>>, []};
+		_ ->
+			generate_all_rows(queue:to_list(Stack), Tree)
+	end.
+
+mark_rebased(#node{ id = RootID } = Node, RootID) ->
+	Node#node{ is_rebased = true };
+mark_rebased(Node, _RootID) ->
+	Node.
+
+generate_leaf({Data, Note}) ->
+	Hash = hash([hash(Data), hash(note_to_binary(Note))]),
+	#node{
+		id = Hash,
+		type = leaf,
+		data = Data,
+		note = Note,
+		max = Note
+	}.
 
 %% Note: This implementation leaves some duplicates in the tree structure.
 %% The produced trees could be a little smaller if these duplicates were 
 %% not present, but removing them with ar_util:unique takes far too long.
-generate_all_rows([]) ->
-	{<<>>, []};
-generate_all_rows(Leaves) ->
-	generate_all_rows(Leaves, Leaves).
-
 generate_all_rows([RootN], Tree) ->
-	{RootN#node.id, Tree};
+	RootID = RootN#node.id,
+	{RootID, Tree};
 generate_all_rows(Row, Tree) ->
-	NewRow = generate_row(Row),
+	NewRow = generate_row(Row, 0),
 	generate_all_rows(NewRow, NewRow ++ Tree).
 
-generate_row([]) -> [];
-generate_row([Left]) -> [generate_node(Left, empty)];
-generate_row([L, R | Rest]) ->
-	[generate_node(L, R) | generate_row(Rest)].
-
-generate_node(Left, empty) ->
-	Left;
-generate_node(L, R) ->
-	#node {
-		id = hash([hash(L#node.id), hash(R#node.id), hash(note_to_binary(L#node.max))]),
+generate_row([], _Shift) -> [];
+generate_row([Left], _Shift) -> [Left];
+generate_row([L, R | Rest], Shift) ->
+	{N, Shift2} = generate_node(L, R, Shift),
+	[N | generate_row(Rest, Shift2)].
+
+generate_node(Left, empty, Shift) ->
+	{Left, Shift};
+generate_node(L, R, Shift) ->
+	LMax = L#node.max,
+	LMax2 = case L#node.is_rebased of true -> Shift + LMax; _ -> LMax end,
+	RMax = R#node.max,
+	RMax2 = case R#node.is_rebased of true -> LMax2 + RMax; _ -> RMax end,
```

Comment:
Why is RMax shifted by LMax2 and not Shift?

Reply:
Consider a tree:
```
                          Root
                  /                  \
        Branch1                      Branch2 <- we are here
       /         \                  /        \
Leaf1            Leaf2          Leaf3     Leaf4

Shift      = 2*ChunkSize
LMax       = 3*ChunkSize, when not rebased 
           =   ChunkSize, when rebased
LMax2      = 3*ChunkSize (always)
RMax       = 4*ChunkSize, when not rebased
           =   ChunkSize, when rebased
RMax2      = 4*ChunkSize (always)
```
The tags passed to ar_merkle:generate_tree when no tree is rebased:

```
[{Leaf1Hash, ChunkSize}, {Leaf2Hash, 2*ChunkSize}, {Leaf3Hash, 3*ChunkSize}, {Leaf4Hash, 4*ChunkSize}]
```

when the subtree with Leaf3 is rebased:

```
[{Leaf1Hash, ChunkSize}, {Leaf2Hash, 2*ChunkSize}, [{Leaf3Hash, ChunkSize}], {Leaf4Hash, 4*ChunkSize}]
```

when the subtree with Leaf4 is rebased:

```
[{Leaf1Hash, ChunkSize}, {Leaf2Hash, 2*ChunkSize}, {Leaf3Hash, 3*ChunkSize}, [{Leaf4Hash, ChunkSize}]]
```

Reply:
Ah, okay. I'm still trying to keep the data formats straight. It sounds like LMax2 and RMax2 will always be the non-rebased values? Or maybe that's not the right term - they'll always be the "global" values? I.e. the values after all 0-offset subtrees have been rebased against a shared Root?

Reply:
Yes!

Reply:
Great - thanks!